[danny.ho 조재호] 지하철 서비스 step3 코드리뷰 요청드립니다.

kcoms555 opened the pull request:
안녕하세요, 대니입니다. 이번 페어 프로그래밍은 ian, ellie와 함께하였습니다.
세 명이라 그런지 민주적으로 빠르게 의사결정 할 수 있어서 좋았습니다.

주요 구현 사항
- path/domain/Path, Person, SectionEdge, SubwayGraph - 최단거리 구하고 거리에 따른 값을 구하기 위해 구현한 영역들입니다.

- path/service/PathService - 최단 경로 관련 거리, 요금 서비스를 제공하는 영역입니다. 최초 최단거리를 구할 시, DB의 연결 간선들을 모두 저장을 합니다. DB에 Station 또는 Section 에 **변경사항**이 생길 시, PathService.newlyUpdated() 를 호출하여 **PathService가 저장하고있는 연결 간선이 변경되었음**을 알려줍니다. 그 경우, PathService는 최단경로 구하는 요청이 들어오면 기존 저장된 연결 간선들을 버리고 DB로부터 새롭게 간선들을 저장합니다.
   - 만약, 데이터베이스에 간선 추가가 동시에 두 번 발생하였는데, 마지막 변경 사항이 반영되지 않은 채 간선이 갱신되는 경우를 방지하기 위해 멀티쓰레드 환경에서도 안정적으로 작동하도록 synchronized 블럭을 사용해보았습니다. 이 부분을 올바르게 적용 한 것인지 리뷰 부탁드리겠습니다.

- auth/ui/AuthenticationPrincipalArgumentResolver - @AuthenticationPrincipal 어노테이션이 명시된 경우, 인증된 LoginMember를  반환하고 인증 실패하면 exception을 던집니다. 만약 해당 어노테이션의 인자 require 가 false인 경우 인증되지 않은 사용자도 해당 서비스를 사용할 수 있는 경우라고 생각하고 null을 반환합니다.

그럼 좋은 리뷰 부탁드리겠습니다!

Jjungs7 committed:
feat(subwaygraph): 노선도 최단경로 조회 추가

Jjungs7 committed:
feat(path): 최단경로 조회 웹 컨트롤러

- DB에서 Section에서 전체 노선 불러오는 기능 추가
- Path 서비스 및 컨트롤러 추가

Jjungs7 committed:
feat(auth): 사용자 인증 및 토큰 발급

- 서비스에 등록된 회원이 아닐 경우 UNAUTHORIZED를 반환

Jjungs7 committed:
feat(favorite): 즐겨찾기 기능 추가

- 인증된 회원만 즐겨찾기 기능을 사용할 수 있음
- 단위 테스트 추가

Jjungs7 committed:
feat(요금): 요금 계산 기능 추가

- 사람을 나타내는 Person 클래스 추가
- 경로에 따른 추가 요금 기능 구현

Jjungs7 committed:
refactor(auth): Principal의 예외 처리 방법 변경

- null체크 하는 부분을 Principal에 위임

Commit:
reset: ian의 브랜치에서 가져옴

Commit:
refactor: 메서드 인자, 이름 수정, 인증시 id로 받도록 수정, 형식, 패키지 위치 등 변경

- 인증 토큰에는 email이 담겼지만 이제 고유 id를 담도록 변경
- /members 에 쓰이지 않는 url 경로(id로 조회/제거) 모두 삭제
- db에 member 조회시 id 또는 email와 password 로만 조회 가능하도록 수정
- exceptions 패키지를 [package]/exceptions로 쓰이는 곳으로 패키지 이동
- 자잘한 메서드명, 메서드 인자의 형식 변경

findstar commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -76,17 +79,43 @@ public void addLineStation(Long lineId, SectionRequest request) {
+
+    private void deleteLineAndSet(Long id) {
+        lineDao.deleteById(id);
+        PathService.newlyUpdated();
```
Spring Event 와 같은 형태로 이런 문제를 풀수도 있습니다. 😄 
 https://javacan.tistory.com/entry/Handle-DomainEvent-with-Spring-ApplicationEventPublisher-EventListener-TransactionalEventListener

kcoms555 replied:
감사합니다. 알려주신 두 가지 방법 중, 저희는 두 번째 방법으로 풀었지만, 기억해 두었다가 똑같은 상황에는 Spring Event로 풀어봐야겠네요.

findstar commented on src/main/java/subway/path/domain/SectionEdge.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    protected double getWeight() {
+        return distance + (fare / 100000.0);
```
긴 숫자는 언더스코어를 붙인 표현식으로 정의하면 좋습니다. java7 부터 사용 가능해요 😄 
https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html

findstar commented on src/main/java/subway/path/domain/SubwayGraph.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.List;
+
+public class SubwayGraph {
```
도메인 모델을 SubwayGraph 로 분리한 구조 좋습니다 💯

findstar commented on src/main/java/subway/path/domain/Person.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.stream.Stream;
+
+public enum Person {
```
💯

findstar commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,32 @@
+    public ResponseEntity<PathResponse> findShortestPath(@AuthenticationPrincipal(required = false) LoginMember loginMember,
+                                                         @RequestParam("source") long sourceId,
+                                                         @RequestParam("target") long targetId) {
+        Path path = pathService.find(sourceId, targetId);
+        int fare = pathService.calculateFare(path, Person.of(loginMember));
+        return ResponseEntity.ok(PathResponse.of(path, fare));
+    }
```
path 와 person 으로 구분된 요금 계산 구조 좋습니다.

findstar commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -5,25 +5,44 @@
+            if (auth.required()) {
+                throw new UnauthenticatedException("유효하지 않은 사용자입니다");
+            }
+            return null;
```
여기서는 required 에 따라서 구분될 수 있지만
null 을 반환하여 외부에서 null 체크가 빈번해 진다면 Null Object Pttern 을 적용해 볼수도 있습니다. 
나중에 참고해 보세요 😄 
https://johngrib.github.io/wiki/null-object-pattern/

kcoms555 replied:
이번에 이 패턴대로 DefaultMember 를 넣어보면서 구현을 해보았습니다. 싱글톤으로 하는게 메모리측면에서 유리하지만, 간단하게 일반적인 객체 생성하듯이 만들어보았습니다. null 체크를 건너띄고 메서드를 호출할 수 있기에 좋네요!

findstar commented on src/main/java/subway/line/dao/SectionDao.java:
```diff
@@ -34,6 +36,25 @@ public Section insert(Line line, Section section) {
     }
 
+    public List<SectionEdge> getSectionEdges() {
+        String sql = "select UST.id as up_station_id, " +
+                "UST.name as up_station_name, " +
+                "DST.id as down_station_id, " +
+                "DST.name as down_station_name, " +
+                "S.distance as distance, " +
+                "L.extra_fare as extra_fare " +
+                "from SECTION S " +
+                "left outer join STATION UST on S.up_station_id = UST.id " +
+                "left outer join STATION DST on S.down_station_id = DST.id " +
+                "left outer join LINE L on S.line_id = L.id";
```
과제에서는 java8 이라서 사용할 수 없지만 
java 13 부터는 """ 와 같은 문자열 리터럴 기법을 사용할 수 있습니다. 
https://www.jrebel.com/blog/using-text-blocks-in-java-13 

저도 이런 문자열 표기법은 불편하다고 생각 해요 😄

kcoms555 replied:
저도 많이 공감합니다...!

findstar commented on src/main/java/subway/favorite/service/FavoriteService.java:
```diff
@@ -0,0 +1,48 @@
+        this.favoriteDao = favoriteDao;
+    }
+
+    public long insert(long memberId, FavoriteRequest favoriteRequest) {
```
insert 메서드 시그니처가 Request 객체로 되어 있는데 이런 경우 다시 null 체크가 필요합니다. 
Request 객체가 인입되어서 컨트롤러에서 input validtion 을 처리하는 게 더 나은 형태로 보여져요. 
일반적으로 Http request 객체를 취급할때는 
서비스 메서드 호출이 되기전에
1. input validation
2. authentication 을 확인한 상태가 되고 
3. 해당 url 의 authorization 까지 확인하기도 합니다.

kcoms555 replied:
객체의 역할을 명시해주셔서 감사합니다!

findstar commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,78 @@
+        return keyHolder.getKey().longValue();
+    }
+
+    public List<FavoriteResponse> findFavoriteResponsesByMemberId(long memberId) {
```
반환 타입이 FavoriteResponse 인데요. 하나의 응답에서 필요로 하는 정보가 
Favorite 과 Station 이 동시에 있을 때 어떻게 해결하는 것이 좋을까요? 
Response 는 Http Request / Response 사이클에서 사용되기 때문에
Controller => Service => Dao 로 이어지는 연결까지 모두 알고 있는 상태가 됩니다. 
여기서는 Favorite 만 반환하고 Service 에서 다시 Station 을 조합하거나, 
복합된 구조를 표현하는 별도의 DTO 를 만들 수도 있습니다.

findstar commented on src/main/java/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,31 @@
+        this.targetStationId = targetStationId;
+    }
+
+    public long getId() {
+        return id;
+    }
```
getId 메서드가 호출되지 않네요

findstar reviewed (changes requested):
안녕하세요. 제임스 입니다. 리뷰가 늦어서 죄송합니다 😭 
미션 코드 구현을 잘 해주셨네요.  👍 👍
몇가지 사항에 대해서 코멘트 남겼습니다.
한번 고민해보시고 다시 PR 부탁드립니다~! 😄

Commit:
refactor: 1차 리뷰 코멘트 사항 반영

- Null 객체 구조 이용해서 LoginMember가 null인 경우의 객체를 DefaultMember로 구현
- http 컨트롤러에서 null 체크하고 Service로 넘기도록 수정
- FavoriteDao의 메서드 findFavoritesByMemberId가 FavoriteResponse를 반환하지않고, Service에서 Response를 반환하게 수정
- 기타 코멘트 사항 반영

Commit:
misc(favorite...): import 최적화
